### PR TITLE
Enforce EIP-7825 per-tx gas cap on settlement

### LINKED ISSUE #4368: bug: Driver doesn't enforce the EIP-7825 per-transaction gas cap
### Problem
Fusaka per-tx gas cap (EIP-7825, 2²⁴ = 16,777,215) is not respected in the driver when building settlements. Transactions are instead capped a 1/2 block gas limit (60M), which is higher than the per tx gas limit.

### Impact
While I couldn't find any indication that we are running into this issue in production, it's theoretically possible for solvers to submit a solution which uses too much gas. Instead of rejecting it, the driver will submit it for the auction without any chance of ever being settled (mempools will reject it during `/settle`)

### To reproduce
If you can reproduce the behaviour, steps to reproduce:
1. Mock an auction with a solution that uses ~20M gas
2. Request /solve 

### Expected behaviour
Driver rejects solution because of gas limit exceeded. It doesn't return the solution in the `/solve` response.

### Additional context
#4261 added this behaviour for the quoting logic. It was never ported to the solution submission logic.

### PR DESCRIPTION
**Pull request by 0xDevNinja**
# Description

The driver caps a settlement's gas estimate at half the block gas limit (e.g. 60M on a 120M block). Fusaka introduced the EIP-7825 per-transaction gas cap of `2^24 - 1 = 16,777,215` on Mainnet; any tx exceeding this is rejected by the mempool. Without this check a solution above the cap can still be returned by `/solve` and forwarded to `/settle`, where it could never be mined. The cap is already enforced on the quote-verification path (#4261) but was never ported to settlement submission.

Note: EIP-7825 is not strictly followed by all chains, some chains picked different per-tx limits (e.g. Arbitrum One uses 32M), so the driver reuses the existing `tx_gas_limit` config knob (#3780) rather than hardcoding a single value. Operators set the per-chain ceiling — `16,777,215` on Mainnet Fusaka, higher elsewhere — and `Gas::new` enforces `min(block_limit / 2, tx_gas_limit)`.

# Changes

* Plumb the `tx_gas_limit` config value through `infra::blockchain::Ethereum` (new `tx_gas_limit()` accessor).
* In `Gas::new`, cap the per-settlement maximum at `min(block_limit / 2, tx_gas_limit)` so over-sized solutions fail fast through the existing `GasLimitExceeded` error.
* Chains with a block gas limit below `2 * tx_gas_limit` keep the tighter half-block behaviour.
* Add unit tests covering the configured-cap path, the boundary case (asserting both `estimate` and the clamped `limit`), the small-block-limit path, and the non-Fusaka case where the half-block cap binds.

## How to test

```
cargo nextest run -p driver settlement::tests
```

All four new tests pass; existing driver tests are unchanged.

Fixes #4368

### CODE REVIEW

**Comment by 0xDevNinja** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 0xDevNinja** [comment]
recheck

**Review by jmg-duarte — APPROVED** [review]

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 1]
```diff
@@ -467,3 +478,49 @@ impl Gas {
...
+mod tests {
+    use super::*;
+
+    fn gas(value: u64) -> eth::Gas {
+        eth::Gas(eth::U256::from(value))
+    }
+
+    #[test]
+    fn rejects_solution_above_eip_7825_cap() {
+        // 120M block (well above the EIP-7825 cap of 16,777,215). Half the
```
Comment (AryanGodara) [inline crates/driver/src/domain/competition/solution/settlement.rs]: nit: can we rewrite this comment, so it reads more clearly.
it can be structured a little bit better 😅

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 2]
```diff
@@ -428,6 +428,11 @@ pub struct Gas {
 }
 
 impl Gas {
+    /// EIP-7825 per-transaction gas cap (2^24 - 1) introduced in Fusaka.
+    /// Any transaction exceeding this will be rejected by the mempool, so a
+    /// solution requiring more gas can never be settled on chain.
+    pub const EIP_7825_TX_GAS_CAP: u64 = (1 << 24) - 1;
```
Comment (AryanGodara) [inline crates/driver/src/domain/competition/solution/settlement.rs]: Zero callers outside this module today. The PR description says the constant is exposed "so callers and tests share the constant", but the test module is a child of `Gas`'s module and reaches it via `super::*` without needing `pub`.


Change `pub const EIP_7825_TX_GAS_CAP` → `pub(crate) const EIP_7825_TX_GAS_CAP` (or private)
(but, double check yourself)

File: crates/driver/src/domain/competition/solution/settlement.rs
Comment (AryanGodara) [inline crates/driver/src/domain/competition/solution/settlement.rs:1]: **Nit**: Is EIP-7825 active on every chain CoW deploys to, or                                                                                                        should this cap honour the existing `tx_gas_limit` config knob (PR #3780) so chains without Fusaka aren't over-restricted?

ref: https://x.com/dapplion/status/2044032991660298559
I checekd for gnosis and this tweet is 2 weeks old. So maybe (keyword: maybe) some chain doesn't have this 👀

**Review by AryanGodara — COMMENTED** [review]
I think the PR is good to merge, just left some small follow-up questions

File: crates/driver/src/domain/competition/solution/settlement.rs
Comment (jmg-duarte) [reply crates/driver/src/domain/competition/solution/settlement.rs:1]: > Nit: Is EIP-7825 active on every chain CoW deploys to, or should this cap honour the existing tx_gas_limit config knob (PR https://github.com/cowprotocol/services/pull/3780) so chains without Fusaka aren't over-restricted?

🧠 good question

File: crates/driver/src/domain/competition/solution/settlement.rs (hunk 2 above)
Comment (0xDevNinja) [reply crates/driver/src/domain/competition/solution/settlement.rs]: Done in [2e4268b](https://github.com/cowprotocol/services/pull/4371/commits/2e4268be6) — made it private (no external callers, tests reach via `super::*`).

File: crates/driver/src/domain/competition/solution/settlement.rs (hunk 1 above)
Comment (0xDevNinja) [reply crates/driver/src/domain/competition/solution/settlement.rs]: Reworded in [2e4268b](https://github.com/cowprotocol/services/pull/4371/commits/2e4268be6).

File: crates/driver/src/domain/competition/solution/settlement.rs
Comment (0xDevNinja) [reply crates/driver/src/domain/competition/solution/settlement.rs:1]: Good catch. EIP-7825 is L1 Fusaka-only, and CoW deploys to several L2s (Arbitrum, Base, Polygon) where the rule does not apply, so a hardcoded 16,777,215 cap would over-restrict their settlements.

Switched the cap to the existing `tx_gas_limit` config knob in [967f2cd](https://github.com/cowprotocol/services/pull/4371/commits/967f2cdf1) — `Gas::new` now uses `min(half_block, tx_gas_limit)`. Operators set the per-chain value (16,777,215 on Fusaka chains, higher elsewhere); no new config surface.

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 3]
```diff
@@ -467,3 +475,71 @@ impl Gas {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn gas(value: u64) -> eth::Gas {
+        eth::Gas(eth::U256::from(value))
+    }
+
+    /// EIP-7825 per-transaction gas cap (2^24 - 1) introduced in Fusaka.
```
Comment (jmg-duarte) [inline crates/driver/src/domain/competition/solution/settlement.rs]:
```suggestion
    /// EIP-7825 per-transaction gas cap (2^24 - 1) introduced in Mainnet's Fusaka hardfork.
```

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 4]
```diff
@@ -441,7 +445,11 @@ impl Gas {
         // will not exceed the remaining space in the block next and ignore transactions
         // whose gas limit exceed the remaining space (without simulating the actual
         // gas required).
-        let max_gas = eth::Gas(block_limit.0 / eth::U256::from(2));
+        // Additionally cap by the configured per-tx gas limit. Operators set
+        // this per chain (e.g. to EIP-7825's 16,777,215 cap on Fusaka chains)
```
Comment (jmg-duarte) [inline crates/driver/src/domain/competition/solution/settlement.rs]: Thanks to this PR I learned that chains opted for different values, like Arb1 did 32M

```suggestion
        // this per chain (e.g. to EIP-7825's 16,777,215 cap on Mainnet Fusaka)
```

Comment (jmg-duarte) [inline crates/driver/src/domain/competition/solution/settlement.rs]: Thanks to this PR I learned that chains opted for different values, like Arb1 did 32M

```suggestion
        // this per chain (e.g. to EIP-7825's 16,777,215 cap on Mainnet Fusaka)
```

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 5]
```diff
@@ -467,3 +475,71 @@ impl Gas {
...
+    use super::*;
+
+    fn gas(value: u64) -> eth::Gas {
+        eth::Gas(eth::U256::from(value))
+    }
+
+    /// EIP-7825 per-transaction gas cap (2^24 - 1) introduced in Fusaka.
+    /// Used in tests as a representative value for the configurable
+    /// `tx_gas_limit` knob on Fusaka chains.
+    const EIP_7825_TX_GAS_CAP: u64 = (1 << 24) - 1;
```
Comment (jmg-duarte) [inline crates/driver/src/domain/competition/solution/settlement.rs]:
```suggestion
    const EIP_7825_MAINNET_TX_GAS_CAP: u64 = (1 << 24) - 1;
```

**Review by jmg-duarte — CHANGES_REQUESTED** [review]

**Comment by 0xDevNinja** [comment]
@jmg-duarte, Addressed all three suggestions in [20dba82](https://github.com/cowprotocol/services/pull/4371/commits/20dba823e):

- Reworded comment at `Gas::new` to read "EIP-7825's 16,777,215 cap on Mainnet Fusaka".
- Updated test const doc to mention Mainnet's Fusaka hardfork explicitly.
- Renamed test const `EIP_7825_TX_GAS_CAP` → `EIP_7825_MAINNET_TX_GAS_CAP`.

Good to know about Arb1 picking 32M — that's exactly why the cap is now driven by the per-chain `tx_gas_limit` config knob rather than a hardcoded value.

**Review by jmg-duarte — APPROVED** [review]

**Review by AryanGodara — APPROVED** [review]
Nit: description claims hardcoded const `EIP_7825_TX_GAS_CAP` but that is no longer true, as it uses tx_gas_limit plumbing.
Update body to reflect the pivot.

**Review by squadgazzz — COMMENTED** [review]
Two minor follow-ups, neither blocking.

PR description note: the Changes list still says the cap is exposed as `Gas::EIP_7825_TX_GAS_CAP` for callers and tests to share, but the final code keeps the constant private to the test module and the per-tx ceiling now comes from the existing `tx_gas_limit` config knob, not a hardcoded constant. Worth tightening before squash so the commit message reflects what shipped.

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 6]
```diff
@@ -467,3 +475,71 @@ impl Gas {
...
+            other => panic!("unexpected error: {other:?}"),
+        }
+    }
+
+    #[test]
+    fn accepts_solution_at_tx_gas_limit() {
+        let block_limit = gas(120_000_000);
+        let tx_gas_limit = gas(EIP_7825_MAINNET_TX_GAS_CAP);
+        let result = Gas::new(tx_gas_limit, block_limit, tx_gas_limit).unwrap();
+        assert_eq!(result.estimate, tx_gas_limit);
```
Comment (squadgazzz) [inline crates/driver/src/domain/competition/solution/settlement.rs:515]: The boundary test only asserts `estimate`, but `limit` is what flows downstream into `required_balance` and the submitted tx's gas field. With `estimate == tx_gas_limit`, the `2.0×` buffer would compute `estimate_with_buffer = 2 * tx_gas_limit`, and the `min(max_gas, ...)` clamp is what keeps the limit at the cap. Worth locking that in.

```suggestion
        assert_eq!(result.estimate, tx_gas_limit);
        assert_eq!(result.limit, tx_gas_limit);
```

Comment (0xDevNinja) [reply crates/driver/src/domain/competition/solution/settlement.rs:515]: Good catch — applied in [78cbd3b](https://github.com/cowprotocol/services/pull/4371/commits/78cbd3b41). Locks in that the 2x estimate buffer can't leak past the configured cap.

**Comment by 0xDevNinja** [comment]
Updated the PR description to reflect the pivot — the cap now reads from the existing `tx_gas_limit` config knob, no hardcoded constant exposed. Also tightened the `accepts_solution_at_tx_gas_limit` test in [78cbd3b](https://github.com/cowprotocol/services/pull/4371/commits/78cbd3b41) per @squadgazzz's suggestion.

File: crates/driver/src/domain/competition/solution/settlement.rs (hunk 6 above)
Comment (jmg-duarte) [reply crates/driver/src/domain/competition/solution/settlement.rs:515]: Even though @squadgazzz didn't reply, I can see you applied the change. Will resolve this and merge
